feat: Go callout server

pweiber opened the pull request:
Adding a go callout server SDK

### Current examples

add_header_mutation
add_body_mutation
header_immediate_response - Redirect Example
basic_callout_server version - Basic server capabilities
jwt_auth - Using jwt authentication
client

Lessyless commented on callouts/go/extproc/internal/server/callout_server.go:
```diff
@@ -0,0 +1,229 @@
+}
+
+func (s *GRPCCalloutService) HandleRequestHeaders(headers *extproc.HttpHeaders) (*extproc.ProcessingResponse, error) {
+	return &extproc.ProcessingResponse{
```
This file doesn't have a line break after the function declarations, but the previous files did. Could you make it consistent?

pweiber commented:
@Lessyless Thanks for the initial review, I did the changes, let me know if I missed something or need new adjustments

Lessyless commented on callouts/go/extproc/examples/jwt_auth/jwt_auth_test.go:
```diff
@@ -0,0 +1,171 @@
+	// Call the HandleRequestHeaders method
+	_, err := service.HandleRequestHeaders(headers)
+
+	// Check if an error occurred
+	if err == nil {
+		t.Fatal("HandleRequestHeaders() did not return an error, want PermissionDenied error")
+	}
+
+	// Check if the error is a permission denied error
+	if got, want := err.Error(), "PermissionDenied"; !strings.Contains(got, want) {
+		t.Errorf("HandleRequestHeaders() error = %v, want %v", got, want)
+	}
+
+	// Check if the error message contains "Authorization token is invalid"
+	if got, want := err.Error(), "Authorization token is invalid"; !strings.Contains(got, want) {
+		t.Errorf("HandleRequestHeaders() error = %v, want %v", got, want)
+	}
```
I would suggest something similar to the following:
```
wantErr := status.Errorf(codes.PermissionDenied, "Authorization token is invalid")
if !cmp.Equal(err, wantErr) {
  // ...
}
```

Lessyless replied:
This comment seems to still be relevant.

Lessyless reviewed (commented):
Looks nice!

pweiber commented:
@jstraceski Should be good now, let me know if you find anything else

@Lessyless I also changed the tests from add_body and add_header to use the cmp, let me know if the cmp and the table-driven are looking good now

Lessyless commented on callouts/go/extproc/examples/client/client.go:
```diff
@@ -0,0 +1,197 @@
+	client := extproc.NewExternalProcessorClient(conn)
+
+	// Unmarshal the JSON string into a slice of ProcessingRequest
+	var rawRequests []map[string]interface{}
+	if err := json.Unmarshal([]byte(jsonData), &rawRequests); err != nil {
+		return nil, err
+	}
+
+	var requests []*extproc.ProcessingRequest
+	for _, rawReq := range rawRequests {
+		reqBytes, err := json.Marshal(rawReq)
+		if err != nil {
+			return nil, err
+		}
+		req := &extproc.ProcessingRequest{}
+		if err := protojson.Unmarshal(reqBytes, req); err != nil {
+			return nil, err
+		}
+		requests = append(requests, req)
+	}
```
I wonder if we can replace this piece of code with the following?
```
	var rawRequests []json.RawMessage
	if err := json.Unmarshal([]byte(jsonData), &rawRequests); err != nil {
		return nil, err
	}

	var requests []*extproc.ProcessingRequest
	for _, rawReq := range rawRequests {
		req := &extproc.ProcessingRequest{}
		if err := protojson.Unmarshal(rawReq, req); err != nil {
			return nil, err
		}
		requests = append(requests, req)
	}
```

Lessyless commented on callouts/go/extproc/examples/client/client.go:
```diff
@@ -0,0 +1,197 @@
+	}
+
+	for _, resp := range responses {
+		respJSON, err := protojson.Marshal(resp)
+		if err != nil {
+			log.Fatalf("Failed to marshal response to JSON: %v", err)
+		}
+		log.Printf("Response: %s", string(respJSON))
```
Did I understand correctly that the purpose of this part is to print responses pretty? If yes, could we use [`prototext.Format`](https://pkg.go.dev/google.golang.org/protobuf/encoding/prototext)?

Lessyless commented on callouts/go/extproc/examples/client/client.go:
```diff
@@ -0,0 +1,197 @@
+	}
+}
+
+var (
+	tls        = flag.Bool("tls", false, "Connection uses TLS if true, else plain TCP")
+	certFile   = flag.String("cert_file", "", "The file containing the CA root cert file")
+	serverAddr = flag.String("addr", "localhost:8181", "The server address in the format of host:port")
+	dataJSON   = flag.String("data", "", "The JSON string containing the ProcessingRequest data")
+)
```
Could you move this group under the imports section (https://google.github.io/styleguide/go/decisions.html#flags).

Lessyless reviewed (commented):
Thank you for the update! Sorry for the confusion, but I meant something a bit different about using `cmp.Diff`. Please see the comments below.

I also took a look at the newly added `client.go` and `client_test.go` files. Just a couple of comments there. Overall good!

pweiber commented:
> Thank you for the update! Sorry for the confusion, but I meant something a bit different about using `cmp.Diff`. Please see the comments below.
> 
> I also took a look at the newly added `client.go` and `client_test.go` files. Just a couple of comments there. Overall good!

@Lessyless I believe it should be good now, let me know if this is what you expected for the `cmp.Diff` and `protocmp`.

For the jwt I had to make some changes based on the errors I was getting, changed the 'iat' and 'exp' to avoid scientific values, I was getting something like:

> s{header:{key:"decoded-exp"  raw_value:"1.720022207e+09"}},
> s{header:{key:"decoded-iat"  raw_value:"1.720018607e+09"}},

Also added the sortHeaders because the headers were showing in a different order than expected

Lessyless reviewed (commented):
Good! Please address the last part of comments, and that's all from my side.

pweiber commented on callouts/go/extproc/examples/jwt_auth/jwt_auth_test.go:
```diff
@@ -38,6 +25,13 @@ func generateTestJWTToken(privateKey []byte, claims jwt.MapClaims) (string, erro
 	return token.SignedString(privateKeyParsed)
 }
 
+// sortHeaders sorts headers in-place by their keys.
+func sortHeaders(headers []*base.HeaderValueOption) {
+	sort.Slice(headers, func(i, j int) bool {
+		return headers[i].Header.Key < headers[j].Header.Key
+	})
+}
+
```
I tried some alternatives to use the `cmpopts.SortSlices` but I'm not sure if that would be the right approach, could you check If that makes sense? If I put the wantResponse and response directly it does not work, had to create the "Sorted"

```go
opts := []cmp.Option{
	protocmp.Transform(),
	cmpopts.SortSlices(func(x, y *base.HeaderValueOption) bool {
		return x.Header.Key < y.Header.Key
	}),
}

wantResponseSorted := wantResponse.Response.(*extproc.ProcessingResponse_RequestHeaders).RequestHeaders.Response.HeaderMutation.SetHeaders
responseSorted := response.Response.(*extproc.ProcessingResponse_RequestHeaders).RequestHeaders.Response.HeaderMutation.SetHeaders

if diff := cmp.Diff(wantResponseSorted, responseSorted, opts...); diff != "" {
	t.Errorf("HandleRequestHeaders() mismatch (-want +got):\n%s", diff)
}
```

Lessyless reviewed (approved):
Approved!
